[NL2ESQL] Adds max function calls allowed to prevent tool being called recursively

qn895 opened the pull request:
Follow up of https://github.com/elastic/kibana/pull/231217. This PR fixes https://github.com/elastic/kibana/issues/229979 and implements a mechanism to stop the nl2esql tool being called indefinitely, and to stop after 5 tries maximum.

<img width="776" height="1193" alt="image" src="https://github.com/user-attachments/assets/69d16eb3-bb0a-45b8-8513-7b39edea448a" />

https://35-187-109-62.sslip.io/projects/UHJvamVjdDoxMTAy/traces/e035f0e5f983d2e61c1153f6f41abd80?selected=&selectedSpanNodeId=U3Bhbjo2NjczNjE%3D



### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [ ] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

### Identify risks

Does this PR introduce any risks? For example, consider risks like hard to test bugs, performance regression, potential of data loss.

Describe the risk, its severity, and mitigation for each identified risk. Invite stakeholders and evaluate how to proceed before merging.

- [ ] [See some risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx)
- [ ] ...





<!--ONMERGE {"backportTargets":["8.17","8.18","8.19","9.0","9.1"]} ONMERGE-->

qn895 committed:
Add max function calls to function to prevent tool being called over and over again if previous look up not successful

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

qn895 commented:
/ci

qn895 commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/329693)
* Commit: 7c59ff07a54b0ac82453f65042b2a0ce2e3848b1

### Failed CI Steps
* [FTR Configs #10](https://buildkite.com/elastic/kibana-pull-request/builds/329693#0198a912-0bdb-4d4e-83ef-757d0aabd090)
* [FTR Configs #10](https://buildkite.com/elastic/kibana-pull-request/builds/329693#0198a935-73d5-46f6-a831-263dc543f2d6)
* [FTR Configs #27](https://buildkite.com/elastic/kibana-pull-request/builds/329693#0198a912-0be6-4d9a-a4df-b7272d9cf23b)
* [FTR Configs #127](https://buildkite.com/elastic/kibana-pull-request/builds/329693#0198a912-0c3d-4670-9152-27b3e58d203e)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/329693#0198a912-0c3d-4670-9152-27b3e58d203e) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/329693/jobs/0198a912-0c3d-4670-9152-27b3e58d203e/artifacts/0198a92a-426b-4c4c-907b-c39897f84366) FTR Configs #<span></span>127 / Cloud Security Posture Test adding Cloud Security Posture Integrations CSPM AWS CIS_AWS Organization Manual Assume Role CIS_AWS Organization Manual Assume Role Workflow
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/329693#0198a912-0be6-4d9a-a4df-b7272d9cf23b) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/329693/jobs/0198a912-0be6-4d9a-a4df-b7272d9cf23b/artifacts/0198a93c-daf9-410c-9461-a94905de5040) FTR Configs #<span></span>27 / Fleet packages test Automatic agent upgrades "after all" hook for "should take agents marked and ready for retry into account and upgrade them"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/329693#0198a912-0be6-4d9a-a4df-b7272d9cf23b) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/329693/jobs/0198a912-0be6-4d9a-a4df-b7272d9cf23b/artifacts/0198a93c-daf9-47c3-a768-fefd7563caf6) FTR Configs #<span></span>27 / Fleet packages test Automatic agent upgrades should take agents marked and ready for retry into account and upgrade them

cc @qn895

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329693","number":329693,"commit":"7c59ff07a54b0ac82453f65042b2a0ce2e3848b1"}],"number":329693}
buildkite-pr-comment-->

qn895 commented:
@elasticmachine merge upstream

qn895 committed:
Add check for previous inferenceIds to update

qn895 committed:
Set toolChoice to non

qn895 commented on x-pack/platform/plugins/shared/inference/server/tasks/nl_to_esql/actions/generate_esql.ts:
```diff
@@ -151,14 +161,27 @@ export const generateEsqlTask = <TToolOptions extends ToolOptions>({
+                    ', '
+                  )} and functions ${functions?.join(
+                    ', '
+                  )}. Try to answer the user's question without using currently available information.`,
```
Thanks for catching that! Updated here [`ac866fa` (#231719)](https://github.com/elastic/kibana/pull/231719/commits/ac866fa67659ff4d46d8c4f88da219108d28dd5d)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/330901)
* Commit: 5097b102b140b07773afeb0954cf1f23270ed492

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/330901#0198c31b-75c5-4d0c-b435-1bde2ed3cc2d)

### History
* :broken_heart: [Build #330116](https://buildkite.com/elastic/kibana-pull-request/builds/330116) failed cd43b33d24648472a35b6cb63881867bd2350c86
* :broken_heart: [Build #330048](https://buildkite.com/elastic/kibana-pull-request/builds/330048) failed 862be7845d0a2f521ac2bfefc4e729251c8f07e6
* :broken_heart: [Build #329693](https://buildkite.com/elastic/kibana-pull-request/builds/329693) failed 7c59ff07a54b0ac82453f65042b2a0ce2e3848b1

cc @qn895

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330901","number":330901,"commit":"5097b102b140b07773afeb0954cf1f23270ed492"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330116","number":330116,"commit":"cd43b33d24648472a35b6cb63881867bd2350c86"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330048","number":330048,"commit":"862be7845d0a2f521ac2bfefc4e729251c8f07e6"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329693","number":329693,"commit":"7c59ff07a54b0ac82453f65042b2a0ce2e3848b1"}],"number":330901}
buildkite-pr-comment-->

qn895 committed:
Fix tools

qn895 commented on x-pack/platform/plugins/shared/inference/server/tasks/nl_to_esql/actions/generate_esql.ts:
```diff
@@ -151,14 +161,27 @@ export const generateEsqlTask = <TToolOptions extends ToolOptions>({
-                },
-              });
+              const args = onlyToolCall.function.arguments;
+              if (args && isPopulatedObject<string, string[]>(args, ['commands', 'functions'])) {
```
Checking args.commands and arg.functions are defined/not null.  
```
* - `isPopulatedObject<keyof MyInterface>({...})`
 *      Limits type to a record with keys of the given interface.
```

qn895 replied:
Updated here [`c2504b8` (#231719)](https://github.com/elastic/kibana/pull/231719/commits/c2504b8e11240297ab6c4e7504b1b99ed190cd10)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/330919)
* Commit: 963551d298033f4d1c8493b68b46c3a542776c2e

### Failed CI Steps
* [FTR Configs #26](https://buildkite.com/elastic/kibana-pull-request/builds/330919#0198c364-f20b-4687-82e0-13a1a9da5845)
* [FTR Configs #85](https://buildkite.com/elastic/kibana-pull-request/builds/330919#0198c364-f242-49f8-863f-d45059a56a2f)
* [FTR Configs #85](https://buildkite.com/elastic/kibana-pull-request/builds/330919#0198c393-9cda-4b07-ae76-491aa84845f6)
* [FTR Configs #108](https://buildkite.com/elastic/kibana-pull-request/builds/330919#0198c364-f258-410e-80ee-186d7c6043eb)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/330919#0198c364-f20b-4687-82e0-13a1a9da5845) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/330919/jobs/0198c364-f20b-4687-82e0-13a1a9da5845/artifacts/0198c392-929f-4bec-a689-e77feedbac1e) FTR Configs #<span></span>26 / apis Streams Endpoints Processing Simulation Successful simulations should gracefully return mappings simulation errors
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/330919#0198c364-f242-49f8-863f-d45059a56a2f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/330919/jobs/0198c364-f242-49f8-863f-d45059a56a2f/artifacts/0198c393-6edf-4bbf-90b9-45e20c2a80ba) FTR Configs #<span></span>85 / Dataset quality API tests data_streams/stats.spec.ts basic Api Key privileges check when required privileges are set and categorized datastreams "after all" hook for "returns creation date when specified"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/330919#0198c393-9cda-4b07-ae76-491aa84845f6) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/330919/jobs/0198c393-9cda-4b07-ae76-491aa84845f6/artifacts/0198c39d-8038-4994-aeb9-6681367e235a) FTR Configs #<span></span>85 / Dataset quality API tests data_streams/stats.spec.ts basic Api Key privileges check when required privileges are set and categorized datastreams "after all" hook for "returns creation date when specified"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/330919#0198c364-f242-49f8-863f-d45059a56a2f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/330919/jobs/0198c364-f242-49f8-863f-d45059a56a2f/artifacts/0198c393-6edf-4f13-a68b-a2743fb08a68) FTR Configs #<span></span>85 / Dataset quality API tests data_streams/stats.spec.ts basic Api Key privileges check when required privileges are set and categorized datastreams "before all" hook for "returns stats correctly"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/330919#0198c393-9cda-4b07-ae76-491aa84845f6) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/330919/jobs/0198c393-9cda-4b07-ae76-491aa84845f6/artifacts/0198c39d-8038-4617-a24b-e79003cbc829) FTR Configs #<span></span>85 / Dataset quality API tests data_streams/stats.spec.ts basic Api Key privileges check when required privileges are set and categorized datastreams "before all" hook for "returns stats correctly"
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/330919#0198c364-f258-410e-80ee-186d7c6043eb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/330919/jobs/0198c364-f258-410e-80ee-186d7c6043eb/artifacts/0198c39c-05e2-4c10-91a5-bd183ccbbae4) FTR Configs #<span></span>108 / Package policies Package Policy - input package behavior should create index templates and update installed_es on package policy creation

### History
* :broken_heart: [Build #330901](https://buildkite.com/elastic/kibana-pull-request/builds/330901) failed 5097b102b140b07773afeb0954cf1f23270ed492
* :broken_heart: [Build #330116](https://buildkite.com/elastic/kibana-pull-request/builds/330116) failed cd43b33d24648472a35b6cb63881867bd2350c86
* :broken_heart: [Build #330048](https://buildkite.com/elastic/kibana-pull-request/builds/330048) failed 862be7845d0a2f521ac2bfefc4e729251c8f07e6
* :broken_heart: [Build #329693](https://buildkite.com/elastic/kibana-pull-request/builds/329693) failed 7c59ff07a54b0ac82453f65042b2a0ce2e3848b1

cc @qn895

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330919","number":330919,"commit":"963551d298033f4d1c8493b68b46c3a542776c2e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330901","number":330901,"commit":"5097b102b140b07773afeb0954cf1f23270ed492"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330116","number":330116,"commit":"cd43b33d24648472a35b6cb63881867bd2350c86"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330048","number":330048,"commit":"862be7845d0a2f521ac2bfefc4e729251c8f07e6"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329693","number":329693,"commit":"7c59ff07a54b0ac82453f65042b2a0ce2e3848b1"}],"number":330919}
buildkite-pr-comment-->

qn895 committed:
Merge upstream/main into branch

qn895 committed:
Merge upstream/main into branch 2

qn895 committed:
Remove isPopulatedObject

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ai-esql-loop

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/331547)
* Commit: 1034ba56da35dfcdb760437d889befee7b5b5551

### Failed CI Steps
* [Build Kibana Distribution](https://buildkite.com/elastic/kibana-pull-request/builds/331547#0198cd87-04b7-4af1-ab6b-71aa2024bbe6)

### History
* :broken_heart: [Build #330919](https://buildkite.com/elastic/kibana-pull-request/builds/330919) failed 963551d298033f4d1c8493b68b46c3a542776c2e
* :broken_heart: [Build #330901](https://buildkite.com/elastic/kibana-pull-request/builds/330901) failed 5097b102b140b07773afeb0954cf1f23270ed492
* :broken_heart: [Build #330116](https://buildkite.com/elastic/kibana-pull-request/builds/330116) failed cd43b33d24648472a35b6cb63881867bd2350c86
* :broken_heart: [Build #330048](https://buildkite.com/elastic/kibana-pull-request/builds/330048) failed 862be7845d0a2f521ac2bfefc4e729251c8f07e6
* :broken_heart: [Build #329693](https://buildkite.com/elastic/kibana-pull-request/builds/329693) failed 7c59ff07a54b0ac82453f65042b2a0ce2e3848b1

cc @qn895

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331547","number":331547,"commit":"1034ba56da35dfcdb760437d889befee7b5b5551"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330919","number":330919,"commit":"963551d298033f4d1c8493b68b46c3a542776c2e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330901","number":330901,"commit":"5097b102b140b07773afeb0954cf1f23270ed492"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330116","number":330116,"commit":"cd43b33d24648472a35b6cb63881867bd2350c86"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330048","number":330048,"commit":"862be7845d0a2f521ac2bfefc4e729251c8f07e6"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/329693","number":329693,"commit":"7c59ff07a54b0ac82453f65042b2a0ce2e3848b1"}],"number":331547}
buildkite-pr-comment-->

dgieselaar reviewed (commented):
@qn895 Can you make sure you address my questions about injecting the tool call and more importantly setting toolChoice/tools in order to give the LLM the chance to produce a final response? Thanks

qn895 committed:
Fix types

qn895 commented on x-pack/platform/plugins/shared/inference/server/tasks/nl_to_esql/actions/generate_esql.ts:
```diff
@@ -107,22 +115,24 @@ export const generateEsqlTask = <TToolOptions extends ToolOptions>({
+                toolCallId: fakeRequestDocsToolCall.toolCallId,
+              },
+            ],
         toolChoice,
```
Updated here [`5097b10` (#231719)](https://github.com/elastic/kibana/pull/231719/commits/5097b102b140b07773afeb0954cf1f23270ed492)

qn895 committed:
Fix weird merge

qn895 commented on x-pack/platform/plugins/shared/inference/server/tasks/nl_to_esql/actions/generate_esql.ts:
```diff
@@ -107,22 +115,24 @@ export const generateEsqlTask = <TToolOptions extends ToolOptions>({
-            toolCallId: fakeRequestDocsToolCall.toolCallId,
-          },
-        ],
+        messages: functionLimitReached
```
I set toolChoice to none and to remove it from tools here [`5097b10` (#231719)](https://github.com/elastic/kibana/pull/231719/commits/5097b102b140b07773afeb0954cf1f23270ed492)

qn895 commented:
Hey @dgieselaar, thanks for your patience as I'm working through the PR. This week is our OnWeek so I'm spending my spare time here to work on this PR as I'm aware how important it is for this fix to get in ASAP. Anyway, I addressed your feedback in the comments so let me know.

@viduni94 If you can give this PR another test, that would be amazing.

dgieselaar commented:
@qn895 ah sorry, it's just that I saw commits coming in, but I didn't see the relevant changes. thanks for responding! I do think this is somewhat urgent so I do appreciate that you're having another look.

viduni94 commented:
> @viduni94 If you can give this PR another test, that would be amazing.

Will do

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/331558)
* Commit: 35e79234f8d3fed6296dd02bda1f29b51c65e3c7

### Failed CI Steps
* [FTR Configs #66](https://buildkite.com/elastic/kibana-pull-request/builds/331558#0198cdbe-a6c1-470a-8cd1-de9b0a4f1e62)
* [FTR Configs #100](https://buildkite.com/elastic/kibana-pull-request/builds/331558#0198cdbe-a6db-4fe7-bac3-304b8bd21e3e)
* [Security Solution Gen AI Evaluations #1](https://buildkite.com/elastic/kibana-pull-request/builds/331558#0198cd9a-db11-4bb1-8eab-e03051fdb7d9)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/331558#0198cdbe-a6c1-470a-8cd1-de9b0a4f1e62) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/331558/jobs/0198cdbe-a6c1-470a-8cd1-de9b0a4f1e62/artifacts/0198cde9-835a-4d79-95cd-4925a54106b3) FTR Configs #<span></span>66 / Dataset Quality Dataset quality handles user privileges User can read logs-* User can monitor some data streams Details page shows insufficient privileges warning for underprivileged data stream
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/331558#0198cdbe-a6db-4fe7-bac3-304b8bd21e3e) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/331558/jobs/0198cdbe-a6db-4fe7-bac3-304b8bd21e3e/artifacts/0198cdd6-dbd9-4977-a08d-be746449ba9a) FTR Configs #<span></span>100 / Visualizations - Group 3 lens app - TSVB Open in Lens Metric should show the "Convert to Lens" menu item

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #331547](https://buildkite.com/elastic/kibana-pull-request/builds/331547) failed 1034ba56da35dfcdb760437d889befee7b5b5551
* :broken_heart: [Build #330919](https://buildkite.com/elastic/kibana-pull-request/builds/330919) failed 963551d298033f4d1c8493b68b46c3a542776c2e
* :broken_heart: [Build #330901](https://buildkite.com/elastic/kibana-pull-request/builds/330901) failed 5097b102b140b07773afeb0954cf1f23270ed492
* :broken_heart: [Build #330116](https://buildkite.com/elastic/kibana-pull-request/builds/330116) failed cd43b33d24648472a35b6cb63881867bd2350c86
* :broken_heart: [Build #330048](https://buildkite.com/elastic/kibana-pull-request/builds/330048) failed 862be7845d0a2f521ac2bfefc4e729251c8f07e6

cc @qn895

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331558","number":331558,"commit":"35e79234f8d3fed6296dd02bda1f29b51c65e3c7"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331547","number":331547,"commit":"1034ba56da35dfcdb760437d889befee7b5b5551"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330919","number":330919,"commit":"963551d298033f4d1c8493b68b46c3a542776c2e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330901","number":330901,"commit":"5097b102b140b07773afeb0954cf1f23270ed492"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330116","number":330116,"commit":"cd43b33d24648472a35b6cb63881867bd2350c86"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330048","number":330048,"commit":"862be7845d0a2f521ac2bfefc4e729251c8f07e6"}],"number":331558}
buildkite-pr-comment-->

viduni94 commented on api_docs/gen_ai_settings.mdx:
```diff
@@ -41,4 +41,3 @@ Contact [@elastic/appex-ai-infra](https://github.com/orgs/elastic/teams/appex-ai
 
 ### Consts, variables and types
 <DocDefinitionList data={genAiSettingsObj.server.misc}/>
-
```
I assume this wasn't intentional. Can we revert it to not have unrelated changes in this PR?

qn895 replied:
Not sure what's happening with git on this one but this current file is identical to main's https://raw.githubusercontent.com/qn895/kibana/b0a69393f9299d6b4ba8d9e793562f57bc142a96/api_docs/gen_ai_settings.mdx

viduni94 reviewed (commented):
Tested the PR and I can't reproduce the issue any more. Looks okay on my end.

I noticed something odd in the tool response:
```
{
  "data": {
    "keywords": [
      "FROM logs-synth-default",
      "LOOKUP JOIN .siem-signals-default ON source.ip",
      "KEEP source.ip, signal.rule.name"
    ],
    "requestedDocumentation": {
      "FROMLOGS-SYNTH-DEFAULT": "\n  ## FROMLOGS-SYNTH-DEFAULT\n\n  There is no FROMLOGS-SYNTH-DEFAULT function or command in ES|QL. Do NOT use it.\n  ",
      "LOOKUP JOIN .siem-signals-default ON source.ip": "\n  ## LOOKUP JOIN .siem-signals-default ON source.ip\n\n  There is no LOOKUP JOIN .siem-signals-default ON source.ip function or command in ES|QL. Do NOT use it.\n  ",
      "KEEPSOURCE.IP,SIGNAL.RULE.NAME": "\n  ## KEEPSOURCE.IP,SIGNAL.RULE.NAME\n\n  There is no KEEPSOURCE.IP,SIGNAL.RULE.NAME function or command in ES|QL. Do NOT use it.\n  ",
      "SYNTAX": "```markdown\n# ES|QL Syntax Reference\n\n## Overview
```

I'm not sure why `"FROMLOGS-SYNTH-DEFAULT": "\n  ## FROMLOGS-SYNTH-DEFAULT\n\n  There is no FROMLOGS-SYNTH-DEFAULT function or command in ES|QL. Do NOT use it.` is returned. It could be a different issue (and maybe out of scope for this PR) But wanted to flag it.

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ai-esql-loop

qn895 committed:
Remove role

qn895 committed:
Merge remote-tracking branch 'upstream/main' into ai-esql-loop

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/332385)
* Commit: 8628f89d1bfd21a318052bbe65b73a5682c1e39f

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :yellow_heart: [Build #331890](https://buildkite.com/elastic/kibana-pull-request/builds/331890) was flaky b0a69393f9299d6b4ba8d9e793562f57bc142a96
* :yellow_heart: [Build #331558](https://buildkite.com/elastic/kibana-pull-request/builds/331558) was flaky 35e79234f8d3fed6296dd02bda1f29b51c65e3c7
* :broken_heart: [Build #331547](https://buildkite.com/elastic/kibana-pull-request/builds/331547) failed 1034ba56da35dfcdb760437d889befee7b5b5551
* :broken_heart: [Build #330919](https://buildkite.com/elastic/kibana-pull-request/builds/330919) failed 963551d298033f4d1c8493b68b46c3a542776c2e
* :broken_heart: [Build #330901](https://buildkite.com/elastic/kibana-pull-request/builds/330901) failed 5097b102b140b07773afeb0954cf1f23270ed492
* :broken_heart: [Build #330116](https://buildkite.com/elastic/kibana-pull-request/builds/330116) failed cd43b33d24648472a35b6cb63881867bd2350c86

cc @qn895

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/332385","number":332385,"commit":"8628f89d1bfd21a318052bbe65b73a5682c1e39f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331890","number":331890,"commit":"b0a69393f9299d6b4ba8d9e793562f57bc142a96"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331558","number":331558,"commit":"35e79234f8d3fed6296dd02bda1f29b51c65e3c7"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/331547","number":331547,"commit":"1034ba56da35dfcdb760437d889befee7b5b5551"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330919","number":330919,"commit":"963551d298033f4d1c8493b68b46c3a542776c2e"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/330901","number":330901,"commit":"5097b102b140b07773afeb0954cf1f23270ed492"}],"number":332385}
buildkite-pr-comment-->

dgieselaar commented on x-pack/platform/plugins/shared/inference/server/tasks/nl_to_esql/actions/generate_esql.ts:
```diff
@@ -150,18 +160,32 @@ export const generateEsqlTask = <TToolOptions extends ToolOptions>({
+              }
+
+              const args =
+                'arguments' in onlyToolCall.function ? onlyToolCall.function.arguments : undefined;
```
we probably need some better type defaults here to prevent having to do this much work to get some props from an object, but we can figure that out later

dgieselaar reviewed (approved):
Thanks @qn895 !

kibanamachine commented:
Starting backport for target branches: 8.17, 8.18, 8.19, 9.0, 9.1

https://github.com/elastic/kibana/actions/runs/17248590364

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.17|Backport failed because of merge conflicts|
|❌|8.18|Backport failed because of merge conflicts|
|❌|8.19|Backport failed because of merge conflicts|
|❌|9.0|Backport failed because of merge conflicts|
|❌|9.1|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 231719
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

qn895 commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.18|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/233153">](https://github.com/elastic/kibana/pull/233153)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.0.1 -->
